backends/tc/backend.cpp

```diff
@@ -40,6 +43,1771 @@ const cstring pnaParserMeta = "pna_main_parser_input_metadata_t"_cs;
+ *  hand, we'd have trouble because the RHS may not have an address
+ *  (from a C perspective, that is; it usually will be stored in memory
+ *  and thus will have an address at the machine-language level) and
+ *  thus may not be a suitable argument to unary &.  Making this a
+ *  function forces the compiler to create a temporary as necessary.
+ */
+static void gen_assign(EBPF::CodeBuilder *bld, const WidthRec *wr) {
+    unsigned int w;
+
+    assert(wr->type == WR_ASSIGN);
+    w = wr->u.assign.w;
+    bld->newline();
+    bld->appendFormat("static __always_inline void assign_%u(u8 *lhs, ", w);
+    append_type_for_width(bld, w);
+    bld->append(" rhs)\n");
+    bld->append("{\n");
+    bld->appendFormat(" __builtin_memcpy(lhs,&rhs.bits[0],%u);\n", (w + 7) >> 3);
+    bld->append("}\n");
+}
+
+void ScanWidths::gen_h(EBPF::CodeBuilder *bld) const {
+    int i;
+
+    bld->appendLine("#define BITS(v) (v).bits");
+    bld->appendLine("#define SETGUARDS(x) do ; while (0)");
+    for (i = 0; i < nwr; i++) {
+        switch (wrv[i].type) {
+            case WR_VALUE:
+                bld->newline();
+                bld->appendFormat("struct internal_bit_%d {\n", wrv[i].u.value);
+                bld->appendFormat("  u8 bits[%d];\n", (wrv[i].u.value + 7) >> 3);
+                bld->append("};\n");
+                bld->newline();
+                bld->appendFormat(
+                    "static __always_inline struct internal_bit_%u loadfrom_%u(u8 *in)\n",
+                    wrv[i].u.value, wrv[i].u.value);
+                bld->append("{\n");
+                bld->appendFormat(" struct internal_bit_%d rv;\n", wrv[i].u.value);
+                bld->append("\n");
+                bld->appendFormat(" __builtin_memcpy(&rv.bits[0],in,%d);\n",
+                                  (wrv[i].u.value + 7) >> 3);
+                if (wrv[i].u.value & 7)
+                    bld->appendFormat(" rv.bits[%d] &= %d;\n", wrv[i].u.value >> 3,
+                                      (1U << (wrv[i].u.value & 7)) - 1);
+                bld->append(" return(rv);\n");
+                bld->append("}\n");
+                break;
+            case WR_CONCAT:
+            case WR_ADD:
+            case WR_SUB:
+            case WR_MUL:
+            case WR_BXSMUL:
+            case WR_CAST:
+            case WR_NEG:
+            case WR_NOT:
+            case WR_SHL_X:
+            case WR_SHL_C:
+            case WR_SHRA_X:
+            case WR_SHRA_C:
+            case WR_SHRL_X:
+            case WR_SHRL_C:
+            case WR_CMP:
+            case WR_BITAND:
+            case WR_BITOR:
+            case WR_BITXOR:
+            case WR_ADDSAT:
+            case WR_SUBSAT:
+            case WR_ASSIGN:
+                break;
+            default:
+                abort();
+                break;
+        }
+    }
+    bld->newline();
+    bld->append(
+        "\
+/*\n\
+ * These are here because EBPF quasi-C does not support aggregate\n\
```

Comment:
The order of this comment seems reversed. It is emitted at the bottom of the file.

Reply:
It was being emitted always, the functions are supposed to be after that
I fixed it so that the comment is only emitted with the functions